### PR TITLE
[build] move hdlc to lib/hdlc

### PR DESCRIPTION
**Pull request by bukepo**
This commit moves hdlc from src/ncp to src/lib/hdlc, src/spinel is also
moved to src/lib/spinel.

### CODE REVIEW

**Comment by bukepo** [comment]
This fixes #4413

File: src/ncp/ncp_base.hpp [hunk 1]
```diff
@@ -51,10 +51,10 @@
...
 #include "common/instance.hpp"
 #include "common/tasklet.hpp"
-#include "spinel/spinel.h"
-#include "spinel/spinel_buffer.hpp"
-#include "spinel/spinel_decoder.hpp"
-#include "spinel/spinel_encoder.hpp"
+#include "lib/spinel/spinel.h"
+#include "lib/spinel/spinel_buffer.hpp"
+#include "lib/spinel/spinel_decoder.hpp"
+#include "lib/spinel/spinel_encoder.hpp"
```
Comment (abtink) [inline src/ncp/ncp_base.hpp]: Can we consider having `src/lib` folder in the include path? This way we won't need `lib/` when calculating the header files.

File: src/lib/hdlc/hdlc.hpp [hunk 2]
```diff
@@ -33,13 +33,12 @@
 #ifndef HDLC_HPP_
 #define HDLC_HPP_
 
-#include "openthread-core-config.h"
```
Comment (abtink) [inline src/lib/hdlc/hdlc.hpp]: Can we consider keeping this? (even though I see that we don't use any configs in hdlc implementation would be good for consistency and future use).
We already include the config header in other libs: e.g. `spinel_encoder.hpp` and `ncp_base.hpp`

Thoughts?

File: src/core/mac/mac_frame.cpp [hunk 3]
```diff
@@ -1031,7 +1031,7 @@ void TxFrame::ProcessTransmitAesCcm(const ExtAddress &aExtAddress)
 
 exit:
     return;
-#endif // OPENTHREAD_MTD || OPENTHREAD_FTD
+#endif // OPENTHREAD_RADIO
```
Comment (abtink) [inline src/core/mac/mac_frame.cpp]: How about moving this to a different PR or commit? Seems very unrelated to rest :)

File: src/ncp/ncp_base.hpp (hunk 1 above)
Comment (bukepo) [reply src/ncp/ncp_base.hpp]: I am trying to follow google's c++ style as a second guide if not covered by OpenThread's style guide. I prefer using a full path because it works better if there are duplicate file names inside a project.
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/lib/hdlc/hdlc.hpp (hunk 2 above)
Comment (bukepo) [reply src/lib/hdlc/hdlc.hpp]: I didn't see spinel_encoder.hpp includes the core config file. I intended to remove dependency on core config, which seems to be focusing on `src/core` config. Also this ease third party utilization of these libraries.

Comment (abtink) [reply src/lib/hdlc/hdlc.hpp]: Yes. You are right. I miss-read `"openthread-spinel-config.h"` in `spinel-encoder.hpp`. I forgot you already helped add a spinel specific config. I agree and think not using core config is definitely the right way for all lib/utils modules.

File: Android.mk [hunk 4]
```diff
@@ -401,7 +401,7 @@ LOCAL_SRC_FILES                            := \
     src/ncp/ncp_base_ftd.cpp                  \
     src/ncp/ncp_base_dispatcher.cpp           \
     src/ncp/ncp_uart.cpp                      \
-    src/spinel/spinel_buffer.cpp              \
+    src/lib/spinel/spinel_buffer.cpp          \
```
Comment (jwhui) [inline Android.mk]: nit: maintain alphabetical order?

File: src/Makefile.am [hunk 5]
```diff
@@ -35,14 +35,14 @@ DIST_SUBDIRS                            = \
     cli                                   \
     ncp                                   \
     posix                                 \
-    spinel                                \
+    lib                                   \
```
Comment (jwhui) [inline src/Makefile.am]: nit: maintain alphabetical order (here and below)
